File: typescript/src/protection-proxy/protection-proxy.test.ts
import { describe, expect, it } from "vitest";
import { ProtectionProxy } from "./protection-proxy.ts";
import { RealSubject } from "./real-subject.ts";

describe("ProtectionProxy", () => {
  it("an admin is forwarded to the real subject", () => {
    expect(new ProtectionProxy(new RealSubject(), "admin").request()).toBe("RealSubject.request()");
  });

  it("a guest is denied and the real subject is never invoked", () => {
    let calls = 0;
    const counting = {
      request: () => {
        calls++;
        return "secret";
      },
    };
    expect(new ProtectionProxy(counting, "guest").request()).toBe("access denied by ProtectionProxy");
    expect(calls).toBe(0);
    expect(new ProtectionProxy(counting, "admin").request()).toBe("secret");
    expect(calls).toBe(1);
  });
});

File: java/src/test/java/com/penapereira/patterns/protectionproxy/ProtectionProxyTest.java
package com.penapereira.patterns.protectionproxy;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

class ProtectionProxyTest {

    @Test
    void anAdminIsForwardedToTheRealSubject() {
        assertEquals("RealSubject.request()", new ProtectionProxy(new RealSubject(), "admin").request());
    }

    @Test
    void aGuestIsDeniedAndTheRealSubjectIsNeverInvoked() {
        AtomicInteger calls = new AtomicInteger();
        Subject counting = () -> {
            calls.incrementAndGet();
            return "secret";
        };
        assertEquals("access denied by ProtectionProxy", new ProtectionProxy(counting, "guest").request());
        assertEquals(0, calls.get());
        assertEquals("secret", new ProtectionProxy(counting, "admin").request());
        assertEquals(1, calls.get());
    }
}
